File: api/convert-heic.js
export const config = { api: { bodyParser: false } };

export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  if (req.method === 'OPTIONS') { res.status(200).end(); return; }
  if (req.method !== 'POST') { res.status(405).end(); return; }

  try {
    // Leer body
    const chunks = [];
    for await (const chunk of req) chunks.push(chunk);
    const buf = Buffer.concat(chunks);
    if (!buf.length) { res.status(400).json({error:'Empty'}); return; }

    // Subir a Cloudinary como data URI y pedir JPG de vuelta
    const b64 = buf.toString('base64');
    const mime = req.headers['content-type'] || 'image/heic';
    const dataUri = `data:${mime};base64,${b64}`;

    const CLOUD = 'dmzkitchen';
    const PRESET = 'heic_convert';

    const body = new URLSearchParams({
      file: dataUri,
      upload_preset: PRESET,
      format: 'jpg',
      quality: '88',
    });

    const upload = await fetch(
      `https://api.cloudinary.com/v1_1/${CLOUD}/image/upload`,
      { method: 'POST', body }
    );

    if (!upload.ok) {
      const err = await upload.text();
      throw new Error('Cloudinary: ' + err.slice(0,200));
    }

    const data = await upload.json();
    const jpgUrl = data.secure_url.replace(/\.[^.]+$/, '.jpg');

    // Descargar el JPG convertido
    const jpgResp = await fetch(jpgUrl);
    if (!jpgResp.ok) throw new Error('Download failed');
    const jpgBuf = Buffer.from(await jpgResp.arrayBuffer());

    res.setHeader('Content-Type', 'image/jpeg');
    res.setHeader('Content-Length', jpgBuf.length);
    res.setHeader('Cache-Control', 'no-store');
    res.status(200).end(jpgBuf);

  } catch(e) {
    console.error('convert-heic:', e.message);
    res.status(500).json({error: e.message});
  }
}

File: api/convert-heic.py
import io
import sys

def handler(request, response):
    """Vercel Python serverless — convierte HEIC/HEVC a JPEG"""
    response.headers["Access-Control-Allow-Origin"] = "*"
    response.headers["Access-Control-Allow-Methods"] = "POST, OPTIONS"
    
    if request.method == "OPTIONS":
        response.status_code = 200
        return
    
    if request.method != "POST":
        response.status_code = 405
        response.write(b'{"error":"Method not allowed"}')
        return
    
    try:
        import pillow_heif
        from PIL import Image
        
        pillow_heif.register_heif_opener()
        
        body = request.body
        if not body:
            response.status_code = 400
            response.write(b'{"error":"Empty body"}')
            return
        
        img = Image.open(io.BytesIO(body))
        out = io.BytesIO()
        img = img.convert("RGB")
        img.save(out, "JPEG", quality=88, optimize=True)
        jpeg_bytes = out.getvalue()
        
        response.headers["Content-Type"] = "image/jpeg"
        response.headers["Content-Length"] = str(len(jpeg_bytes))
        response.headers["Cache-Control"] = "no-store"
        response.status_code = 200
        response.write(jpeg_bytes)
        
    except Exception as e:
        response.status_code = 500
        response.write(f'{{"error":"{str(e)}"}}'.encode())
